internal/providers/oracle/network.go

```diff
@@ -0,0 +1,157 @@
+}
+
+func (g oracleNetwork) ID() string {
+	return g.id
+}
+
+func (g oracleNetwork) Name() string {
+	return g.name
+}
+
+type oracleSubnet struct {
+	cidrs    []string
+	id       string
+	location string
+	name     string
+}
+
+func (g oracleSubnet) CIDRs() []string {
+	return g.cidrs
+}
+
+func (g oracleSubnet) ID() string {
+	return g.id
+}
+
+func (g oracleSubnet) Location() string {
+	return g.location
+}
+
+func (g oracleSubnet) Name() string {
+	return g.name
+}
+
+type oracleRouteTable struct {
+	id   string
+	name string
+}
+
+func (g oracleRouteTable) ID() string {
+	return g.id
+}
+
+func (g oracleRouteTable) Name() string {
+	return g.name
+}
+
+type oracleNetworkService struct {
+	client *oci.VirtualNetwork
+	logger logrus.FieldLogger
+}
+
+// NewNetworkService returns a new Oracle network Service
+func NewNetworkService(secret *secret.SecretItemResponse, logger logrus.FieldLogger) (network.Service, error) {
+	o, err := oci.NewOCI(secretOracle.CreateOCICredential(secret.Values))
+	if err != nil {
+		return nil, emperror.Wrap(err, "failed to create OCI credential")
+	}
+	client, err := o.NewVirtualNetworkClient()
+	if err != nil {
+		return nil, emperror.Wrap(err, "failed to create virtual network client")
+	}
+	return &oracleNetworkService{
+		client: client,
+		logger: logger,
+	}, nil
+}
+
+// ListNetworks returns VCNs
+func (ns *oracleNetworkService) ListNetworks() ([]network.Network, error) {
+	vcns, err := ns.client.GetVCNs()
+	if err != nil {
+		return nil, emperror.Wrap(err, "failed to retreive VCNs")
+	}
+	networks := make([]network.Network, len(vcns))
+	for idx, item := range vcns {
+		networks[idx] = &oracleNetwork{
+			cidrs: []string{deref(item.CidrBlock)},
+			id:    deref(item.Id),
+			name:  deref(item.DisplayName),
+		}
+	}
+	return networks, nil
+}
+
+// ListSubnets returns VCN subnetworks
+func (ns *oracleNetworkService) ListSubnets(networkID string) ([]network.Subnet, error) {
+	sns, err := ns.client.GetSubnets(&networkID)
+	if err != nil {
+		return nil, emperror.WrapWith(err, "failed to retreive subnets", "networkID", networkID)
+	}
+	subnets := make([]network.Subnet, len(sns))
+	for idx, item := range sns {
+		subnets[idx] = &oracleSubnet{
+			cidrs:    []string{deref(item.CidrBlock)},
+			id:       deref(item.Id),
+			location: deref(item.AvailabilityDomain),
+			name:     deref(item.DisplayName),
+		}
+	}
+	return subnets, nil
+}
+
+// ListRouteTables returns VCN route tables
+func (ns *oracleNetworkService) ListRouteTables(networkID string) ([]network.RouteTable, error) {
+	rts, err := ns.client.GetRouteTables(&networkID)
+	if err != nil {
+		return nil, emperror.WrapWith(err, "failed to retreive route tables", "networkID", networkID)
+	}
+	routeTables := make([]network.RouteTable, len(rts))
+	for idx, item := range rts {
+		routeTables[idx] = &oracleRouteTable{
+			id:   deref(item.Id),
+			name: deref(item.DisplayName),
+		}
+	}
+	return routeTables, nil
+}
+
+func deref(p *string) string {
```

Comment:
Should this be moved to some common package such as it can be used in other places as well?

Reply:
It could be, but then it'd have to be renamed to something like `deref_string_or_empty` to avoid confusion and collisions and that wouldn't be so succinct anymore.